src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/CertificateRequest.cs

```diff
@@ -530,20 +534,92 @@ public X509Certificate2 CreateSelfSigned(DateTimeOffset notBefore, DateTimeOffse
+                Validity = new ValidityAsn(notBefore, notAfter),
+                Subject = SubjectName.RawData,
             };
 
-            tbsCertificate.Extensions.AddRange(CertificateExtensions);
+            if (CertificateExtensions.Count > 0)
+            {
+                HashSet<string> usedOids = new HashSet<string>(CertificateExtensions.Count);
+                List<X509ExtensionAsn> extensionAsns = new List<X509ExtensionAsn>(CertificateExtensions.Count);
+
+                // An interesting quirk of skipping null values here is that
+                // Extensions.Count == 0 => no extensions
+                // Extensions.ContainsOnly(null) => empty extensions list
+
+                foreach (X509Extension extension in CertificateExtensions)
+                {
+                    if (extension == null)
+                        continue;
+
+                    if (!usedOids.Add(extension.Oid.Value))
+                    {
+                        throw new InvalidOperationException(
+                            SR.Format(SR.Cryptography_CertReq_DuplicateExtension, extension.Oid.Value));
+                    }
+
+                    extensionAsns.Add(new X509ExtensionAsn(extension, false));
+                }
 
-            return new X509Certificate2(tbsCertificate.Sign(generator, HashAlgorithm));
+                tbsCertificate.Extensions = new ExtensionsAsn { Extensions = extensionAsns.ToArray() };
+            }
+
+            using (AsnWriter writer = AsnSerializer.Serialize(tbsCertificate, AsnEncodingRules.DER))
+            {
+                byte[] encodedTbsCertificate = writer.Encode();                
+                CertificateAsn certificate = new CertificateAsn
+                {
+                    TbsCertificate = tbsCertificate,
+                    SignatureAlgorithm = signatureAlgorithmAsn,
+                    SignatureValue = generator.SignData(encodedTbsCertificate, HashAlgorithm)
+                };
+
+                using (AsnWriter signedWriter = AsnSerializer.Serialize(certificate, AsnEncodingRules.DER))
+                {
+                    return new X509Certificate2(signedWriter.Encode());
+                }
+            }
+        }
+
+        private ReadOnlyMemory<byte> NormalizeSerialNumber(byte[] serialNumber)
+        {
+            if (serialNumber[0] >= 0x80)
+            {
+                var newSerialNumber = new byte[serialNumber.Length + 1];
+                newSerialNumber[0] = 0;
+                serialNumber.CopyTo(newSerialNumber, 1);
+                return newSerialNumber;
+            }
+            else
+            {
+                int leadingZeros = 0;
+                while (leadingZeros < serialNumber.Length - 1 && serialNumber[leadingZeros] == 0)
+                {
+                    leadingZeros++;
+                }
+                return new ReadOnlyMemory<byte>(serialNumber, leadingZeros, serialNumber.Length - leadingZeros);
+            }
         }
```

Comment:
This whole methods feels ugly as hell. I thought about declaring the SerialNumber as `BigInteger`, but that only has little-endian constructor and methods, so it feels just as cumbersome.

Reply:
It has unsigned big-endian ctors and methods in the netcoreapp target.  But this does the work and avoids the extra reading, copying, and writing that BigInteger does.

Since this normalization is a promise of this class (vs rejecting malformed data) this is good.  I prefer this to using BigInteger for the Asn.

Reply:
```C#
// Did this already check for non-null and non-empty elsewhere?
if (serialNumber[0] == 0 || serialNumber[0] >= 0x80)
{
    // Read it unsigned to stay positive(, man).
    BigInteger serialValue = new BigInteger(serialNumber, isUnsigned: true, isBigEndian: true);
    // Write it signed, so it adds a padding byte if needed
    serialNumber = serialValue.ToByteArray(isUnsigned: false, isBigEndian: true);
}
```

Is the short form (for builds which can use the new BigInteger API).  It's just more expensive (always creates a new output array, creates an intermediate `uint[]` if the parsed value wouldn't fit in an `int`).  Granted, this code isn't that perf-sensitive.

Reply:
Good catch with the negative number. I will write unit test for it.